larpandoracontent/LArControlFlow/NeutrinoIdTool.cc

```diff
@@ -10,30 +10,459 @@
+        LArPcaHelper::RunPca(pointsInSphere, centroid, eigenValues, eigenVectors);
+
+
+        float nuNFinalStatePfos(static_cast<float>(nuFinalStates.size()));
+        float nuVertexY(nuVertex.GetY());
+        float nuWeightedDirZ(nuWeightedDir.GetZ());
+        float nuNSpacePointsInSphere(static_cast<float>(pointsInSphere.size()));
+        float nuEigenRatioInSphere(eigenValues.GetY() / eigenValues.GetX());
+
+        // Cosmic-ray features
+        unsigned int nCRHitsMax(0);
+        unsigned int nCRHitsTotal(0);
+        float crLongestTrackDirY(std::numeric_limits<float>::max());
+        float crLongestTrackDeflection(-std::numeric_limits<float>::max());
+
+        for (const ParticleFlowObject *const pPfo : crPfos)
+        {
+            CartesianPointVector spacePoints;                                                                                                    
+            this->GetSpacePoints(pPfo, spacePoints);
+            
+            nCRHitsTotal += spacePoints.size();
+            
+            if (spacePoints.size() < 5) continue;
+            
+            if (spacePoints.size() > nCRHitsMax)
+            {
+                nCRHitsMax = spacePoints.size();
+                const CartesianVector upperDir(this->GetUpperDirection(spacePoints));
+                const CartesianVector lowerDir(this->GetLowerDirection(spacePoints));
+
+                crLongestTrackDirY = upperDir.GetY();
+                crLongestTrackDeflection = 1 - upperDir.GetDotProduct(lowerDir * (-1));
+            }
+        }
+
+        if (nCRHitsMax == 0) return;
+        if (nCRHitsTotal == 0) return;
+
+        float crFracHitsInLongestTrack = static_cast<float>(nCRHitsMax)/static_cast<float>(nCRHitsTotal);
+
+        // Push the features to the feature vector
+        m_featureVector.push_back(nuNFinalStatePfos);
+        m_featureVector.push_back(nuNHitsTotal);
+        m_featureVector.push_back(nuVertexY);
+        m_featureVector.push_back(nuWeightedDirZ);
+        m_featureVector.push_back(nuNSpacePointsInSphere);
+        m_featureVector.push_back(nuEigenRatioInSphere);
+        m_featureVector.push_back(crLongestTrackDirY);
+        m_featureVector.push_back(crLongestTrackDeflection);
+        m_featureVector.push_back(crFracHitsInLongestTrack);
+        m_featureVector.push_back(nCRHitsMax);
+
+        m_isAvailable = true;
+    }
+    catch (StatusCodeException &)
+    {
+        return;
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool NeutrinoIdTool::SliceFeatures::IsFeatureVectorAvailable() const
+{
+    return m_isAvailable;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+        
+void NeutrinoIdTool::SliceFeatures::GetFeatureVector(SupportVectorMachine::DoubleVector &featureVector) const
+{
+    if (!m_isAvailable)
```

Comment:
Do we also need to check whether the feature vector is empty, or would that be OK?

Reply:
The feature vectors all must have the same size and the order of the features within them must be the same. The idea is we will only get out of the SliceFeatures constructor with `m_isAvailable = true` if all zero check were passed an no exceptions were thrown. We then never touch `m_featureVector` again, all member functions are const. 

https://github.com/a-d-smith/LArContent/blob/8feae693de59e1f2d06b15f8a308fd65d55c54e0/larpandoracontent/LArControlFlow/NeutrinoIdTool.cc#L326-L338

Reply:
The feature vectors all must have the same size and the order of the features within them must be the same. The idea is we will only get out of the SliceFeatures constructor with `m_isAvailable = true` if all zero check were passed an no exceptions were thrown. We then never touch `m_featureVector` again, all member functions are const. 

https://github.com/a-d-smith/LArContent/blob/8feae693de59e1f2d06b15f8a308fd65d55c54e0/larpandoracontent/LArControlFlow/NeutrinoIdTool.cc#L326-L338